index.js

```diff
@@ -70,48 +95,112 @@ module.exports = async function(options) {
+        if (processes <= 0) {
+          processes = cpus().length + processes;
         }
-      }
-      console.log(`Starting ${processes} cluster child processes${capped}`);
-      for (let i = 0; i < processes; i++) {
-        clusterFork();
-      }
-      cluster.on('exit', (worker, code, signal) => {
-        if (code !== 0) {
-          if ((Date.now() - worker.bornAt) < guardTime) {
-            console.error(`Worker process ${worker.process.pid} failed in ${seconds(Date.now() - worker.bornAt)}, waiting ${seconds(guardTime)} before restart`);
-            setTimeout(() => {
-              respawn(worker);
-            }, guardTime);
+        let capped = '';
+        if (processes > cpus().length) {
+          processes = cpus().length;
+          capped = ' (capped to number of CPU cores)';
+        }
+        if (processes < 2) {
+          processes = 2;
+          if (capped) {
+            capped = ' (less than 2 cores, capped to minimum of 2)';
           } else {
-            respawn(worker);
+            capped = ' (using minimum of 2)';
+          }
+        }
+        console.log(`Starting ${processes} cluster child processes${capped}`);
+        for (let i = 0; i < processes; i++) {
+          clusterFork();
+        }
+        cluster.on('exit', (worker, code, signal) => {
+          if (code !== 0) {
+            if ((Date.now() - worker.bornAt) < guardTime) {
+              console.error(`Worker process ${worker.process.pid} failed in ${seconds(Date.now() - worker.bornAt)}, waiting ${seconds(guardTime)} before restart`);
+              setTimeout(() => {
+                respawn(worker);
+              }, guardTime);
+            } else {
+              respawn(worker);
+            }
           }
+        });
+        span.end();
+        if (options.openTelemetrySDK) {
+          await options.openTelemetrySDK
+            .shutdown()
+            .then(() => console.log('OpenTelemetry stopped.'));
         }
+        return null;
       });
-      return null;
     } else {
+      // continue as a worker operation, the pid should be recorded by the auto instrumentation
+      spanName += ':worker';
       console.log(`Cluster worker ${process.pid} started`);
     }
   }
 
+  // Create and activate the root span for the boot tracer
+  const self = await telemetry.aposStartActiveSpan(spanName, async (span) => {
+    try {
+      const res = await apostrophe(options, telemetry);
+      if (!res.taskRan) {
+        // Tasks exit early, pass the root span in order to not break
+        // the task handler. Run event is lost, no way out when early exits exist.
+        await res.emit('run', res.isTask(), span);
+      }
+      span.setStatus(telemetry.SpanStatusCode.OK);
+      return res;
+    } catch (err) {
+      telemetry.aposHandleError(span, err);
+      error = err;
+    } finally {
+      span.end();
+    }
+  });
+
+  if (self && self.taskRan && options.openTelemetrySDK) {
```

Comment:
This can go away if we factor the process.exit() calls to `apos._exit()` and allow passing in a `beforeExit` handler.